File: Java/_1406.java
package Algorithm.BaekJoon;

import java.io.*;
import java.util.*;

// 스택
public class _1406 {

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        String str = br.readLine();

        Stack<Character> left = new Stack<>();
        Stack<Character> right = new Stack<>();

        for(Character c : str.toCharArray()){
            left.push(c);
        }

        int N = Integer.parseInt(br.readLine());

        while(N-- > 0){

            String[] line = br.readLine().split(" ");
            String c = line[0];

            if(c.equals("L")){
                if(left.isEmpty()){
                    continue;
                } else {
                    right.push(left.pop());
                }
            } else if(c.equals("D")){
                if(right.isEmpty()){
                    continue;
                } else{
                    left.push(right.pop());
                }
            } else if(c.equals("B")){
                if(left.isEmpty()){
                    continue;
                } else{
                    left.pop();
                }
            } else if(c.equals("P")){
                left.push(line[1].charAt(0));
            }
        }

        while(!left.isEmpty()){
            right.push(left.pop());
        }

        StringBuilder sb = new StringBuilder();
        while(!right.isEmpty()){
            sb.append(right.pop());
        }

        System.out.println(sb);
    }
}

File: Python/1406.py
import sys

f = lambda : sys.stdin.readline().rstrip()

left = [i for i in f()]
right = list()

for _ in range(int(f())):
    line = f().split()
    c = line[0]
    if len(line) > 1:
        v = line[1]

    if c == "L":
        if len(left) == 0:
            continue
        else:
            right.append(left.pop())
    elif c == "D":
        if len(right) == 0:
            continue
        else:
            left.append(right.pop())
    elif c == "B":
        if len(left) == 0:
            continue
        else:
            left.pop()
    elif c == "P":
        left.append(v)

while len(left) != 0:
    right.append(left.pop())

while len(right) != 0:
    print(right.pop(), end="")